src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,342 @@
+        }
+
+        /// <summary>
+        /// For tuple literals, we just return the element.
+        /// For expressions with tuple type, we access `Item{i}`.
+        /// </summary>
+        private BoundExpression GetTuplePart(BoundExpression tuple, int i)
+        {
+            // Example:
+            // (1, 2) == (1, 2);
+            if (tuple.Kind == BoundKind.TupleLiteral)
+            {
+                return ((BoundTupleLiteral)tuple).Arguments[i];
+            }
+
+            Debug.Assert(tuple.Type.IsTupleType);
+
+            // Example:
+            // t == GetTuple();
+            // t == ((byte, byte)) (1, 2);
+            // t == ((short, short))((int, int))(1L, 2L);
+            return MakeTupleFieldAccessAndReportUseSiteDiagnostics(tuple, tuple.Syntax, tuple.Type.TupleElements[i]);
+        }
+
+        /// <summary>
+        /// Produce an element-wise comparison and logic to ensure the result is a bool type.
+        ///
+        /// If an element-wise comparison doesn't return bool, then:
+        /// - if it is dynamic, we'll do `!(comparisonResult.false)` or `comparisonResult.true`
+        /// - if it implicitly converts to bool, we'll just do the conversion
+        /// - otherwise, we'll do `!(comparisonResult.false)` or `comparisonResult.true` (as we'd do for `if` or `while`)
+        /// </summary>
+        private BoundExpression RewriteTupleSingleOperator(TupleBinaryOperatorInfo.Single single,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType, BinaryOperatorKind operatorKind)
+        {
+            if (single.Kind.IsDynamic())
+            {
+                // Produce
+                // !((left == right).op_false)
+                // (left != right).op_true
+
+                BoundExpression dynamicResult = _dynamicFactory.MakeDynamicBinaryOperator(single.Kind, left, right, isCompoundAssignment: false, _compilation.DynamicType).ToExpression();
+                if (operatorKind == BinaryOperatorKind.Equal)
+                {
+                    return _factory.Not(MakeUnaryOperator(UnaryOperatorKind.DynamicFalse, left.Syntax, method: null, dynamicResult, boolType));
+                }
+                else
+                {
+                    return MakeUnaryOperator(UnaryOperatorKind.DynamicTrue, left.Syntax, method: null, dynamicResult, boolType);
+                }
+            }
+
+            if (left.IsLiteralNull() && right.IsLiteralNull())
+            {
+                // For `null == null` this is special-cased during initial binding
+                return new BoundLiteral(left.Syntax, ConstantValue.Create(operatorKind == BinaryOperatorKind.Equal), boolType);
+            }
+
+            // PROTOTYPE(tuple-equality) checked
+            // We leave the null literal in nullable-null conversions unconverted because MakeBinaryOperator has special rules for it
+            bool isNullableNullConversion = operatorKind.OperandTypes() != BinaryOperatorKind.NullableNull;
```

Comment:
>isNullableNullConversion [](start = 17, length = 24)

There is still something strange about this statement. The local is named "isNullableNullConversion", but it looks like the initializer checks for an opposite situation. Perhaps the local name is confusing, but the comment above also talks about nullable-null conversions.

Reply:
Sorry about that. There is definitely something wrong. Investigating.

---
In reply to: [172339570](https://github.com/dotnet/roslyn/pull/23356#discussion_r172339570) [](ancestors = 172339570)

Reply:
Thanks a lot for catching that, Aleksey. 
The element-wise binary operator for nullable-null conversion has a kind, but no operator and no converted types. So no conversion is needed for either side (the `null` or the other one) during lowering.
There was a similar bug for nullable-null during binding, where I'd used an error type, when in fact leaving the types as null is perfectly fine.